### ISSUE TITLE
The expr/c syntax class reports errors with incorrect blame

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#1412]
Given the following program:

``` racket
#lang racket
(require (for-syntax syntax/parse))

(define-syntax contracted
  (syntax-parser
    [(_ {~var x (expr/c #'string?)})
     #'x.c]))

(contracted #f)
```

The following contract error is raised:

```
broke its own contract
  promised: string?
  produced: #f
  in: string?
  contract from: /private/tmp/expr-c.rkt
  blaming: /private/tmp/expr-c.rkt
   (assuming the contract is correct)
  at: /private/tmp/expr-c.rkt:9.12
  context...:
   /Users/alexis/gits/racket/racket/racket/collects/racket/contract/private/blame.rkt:159:0: raise-blame-error16
   /private/tmp/expr-c.rkt: [running body]
```

The `broke its own contract` message seems unintuitive, given that the contracted value is provided by the user. Instead, I would expect a standard `contract violation` error message.

**Comment by lexi-lambda** [comment]
After doing a tiny bit of digging, it looks like `expr/c` just defers to `wrap-expr/c` from `syntax/contract`, which produces a syntax object that uses the `contract` form directly. Because of this, the wrapped expression’s value is the positive party, which causes the aforementioned error message.

I personally think that the behavior should be reversed—that is, blame parties should be swapped when using `wrap-expr/c`. However, this is a little bit tricky, because it could be wrong depending on how the macro uses the value. For example, something that expands into a *definition* (that is, something that is semantically similar to `define/contract`) would want the current behavior, but something that uses the expression as a *value* (all the cases I’ve used `expr/c` so far) would want the blame to be swapped.

Therefore, it seems like this is something that should probably be exposed to the user, but I’m not completely sure how to do it. @rfindler and @rmculpepper, do you have any suggestions?

**Comment by rfindler** [comment]
I don't know the right thing here, but a recent paper (mostly written by @chrdimo) captures the right way to think about this kind of problem well, in my opinion:

  http://www.eecs.northwestern.edu/~robby/pubs/papers/icfp2016-dnff.pdf

**Comment by lexi-lambda** [comment]
I did read that paper a couple days ago (and it was very interesting 🙂), and I just looked over it again. I’m not completely sure if you had a specific part of it in mind related to this or if you would just generally recommend it.

I think the key part here, though, is that macros that expand into expressions can act conceptually like functions. The is, in a (somewhat silly) macro like this one:

```racket
(define-syntax-rule (prefix-hello str-expr)
  (string-append "Hello, " str-expr))
```

…the provided expression is used in a way that a user would likely conceptualize as similar to function invocation. Obviously, the above macro would be better as a real function, but some more complicated macros use expressions in similar ways (such as [monadic `do`](http://docs.racket-lang.org/functional/interfaces.html#%28form._%28%28lib._data%2Fmonad..rkt%29._do%29%29)).

Other macros, however, use expressions in ways that are *not* “function-like”:

```racket
(define-syntax-rule (with-strings ([id expr] ...) body ...)
  (local [(define/with-contract id string? expr) ...] body ...))
```

In this case, the contracted expressions are bound to identifiers rather than used as part of an internal expression. An equivalent implementation to the above macro that uses `expr/c` would want the current behavior, not the “function-like” behavior described above.

This isn’t something that can be detected automatically, since it’s impossible to know how a particular expression will be used, but I think the current behavior often misaligns with a user’s conceptual model (or at least it does with mine). I’m just not sure if the behavior should be controlled by yet *another* keyword parameter or if it should be something else (and I’m also not sure what the correct default should be).
